library/ssl_cli.c

```diff
@@ -158,47 +154,62 @@ static void ssl_write_renegotiation_ext( mbedtls_ssl_context *ssl,
  */
 #if defined(MBEDTLS_SSL_PROTO_TLS1_2) && \
     defined(MBEDTLS_KEY_EXCHANGE__WITH_CERT__ENABLED)
-static void ssl_write_signature_algorithms_ext( mbedtls_ssl_context *ssl,
-                                                unsigned char *buf,
-                                                size_t *olen )
+static int ssl_write_signature_algorithms_ext( mbedtls_ssl_context *ssl,
+					       unsigned char *buf,
+					       const unsigned char *end,
+					       size_t *olen )
 {
     unsigned char *p = buf;
-    const unsigned char *end = ssl->out_msg + MBEDTLS_SSL_MAX_CONTENT_LEN;
     size_t sig_alg_len = 0;
     const int *md;
+    
 #if defined(MBEDTLS_RSA_C) || defined(MBEDTLS_ECDSA_C)
     unsigned char *sig_alg_list = buf + 6;
 #endif
+    
+    size_t sz_per_alg = 0;
+#if defined(MBEDTLS_RSA_C)
+    sz_per_alg += 2;
+#endif
+#if defined(MBEDTLS_ECDSA_C)
+    sz_per_alg += 2;
+#endif    
 
     *olen = 0;
 
     if( ssl->conf->max_minor_ver != MBEDTLS_SSL_MINOR_VERSION_3 )
-        return;
+        return( 0 );
 
     MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, adding signature_algorithms extension" ) );
 
-    for( md = ssl->conf->sig_hashes; *md != MBEDTLS_MD_NONE; md++ )
+    if( ssl->conf->sig_hashes == NULL )
+	return( MBEDTLS_ERR_SSL_BAD_CONFIG );
+
+    for( md = ssl->conf->sig_hashes;	     
+	 sig_alg_len <= MBEDTLS_SSL_MAX_SIG_HASH_ALGORITHMS - sz_per_alg && *md != MBEDTLS_MD_NONE;
+	 md++ )
     {
 #if defined(MBEDTLS_ECDSA_C)
-        sig_alg_len += 2;
+	    sig_alg_len += 2;
 #endif
 #if defined(MBEDTLS_RSA_C)
-        sig_alg_len += 2;
+	    sig_alg_len += 2;
 #endif
     }
 
-    if( end < p || (size_t)( end - p ) < sig_alg_len + 6 )
-    {
-        MBEDTLS_SSL_DEBUG_MSG( 1, ( "buffer too small" ) );
-        return;
-    }
+    if( sig_alg_len == 0 )
+	return( 0 );
```

Comment:
As far as I understand the standard, this won't work. Dropping the extension means falling back to the default (ok), but providing an empty list means to support no sig-hash-pairs at all.